### PR TITLE
Linkify config description text.

### PR DESCRIPTION
**Pull request by kadams54**

### CODE REVIEW

**Comment by kadams54** [comment]
## QA

1. Add the Wordpress charm to your canvas.
2. In the inspector, navigate into "Configure".
3. Scroll to the bottom and check the description for `wp-content`; the github link in it should be linkified.

File: jujugui/static/gui/src/app/components/inspector-config/inspector-config.js [hunk 1]
```diff
@@ -21,6 +21,18 @@ with this program.  If not, see <http://www.gnu.org/licenses/>.
 YUI.add('inspector-config', function() {
 
   juju.components.Configuration = React.createClass({
+    propTypes: {
```
Comment (kadams54) [inline jujugui/static/gui/src/app/components/inspector-config/inspector-config.js]: Added all of these as a drive-by cleanup.

File: jujugui/static/gui/src/app/components/inspector/inspector.js [hunk 2]
```diff
@@ -21,6 +21,9 @@ with this program.  If not, see <http://www.gnu.org/licenses/>.
 YUI.add('inspector-component', function() {
 
   juju.components.Inspector = React.createClass({
+    propTypes: {
```
Comment (kadams54) [inline jujugui/static/gui/src/app/components/inspector/inspector.js]: Didn't want to get sucked into adding ALL the propTypes for the `Inspector`. That can wait for another day and a dedicated card.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/3969/
Test PASSed.

File: jujugui/static/gui/src/app/components/boolean-config/boolean-config.js [hunk 3]
```diff
@@ -92,8 +92,8 @@ YUI.add('boolean-config', function() {
               <div className="boolean-config--handle"></div>
             </label>
           </div>
-          <div className="boolean-config--description">
-            {this.props.option.description}
+          <div className="boolean-config--description"
+            dangerouslySetInnerHTML={{__html: this.props.option.description}}>
```
Comment (huwshimi) [inline jujugui/static/gui/src/app/components/boolean-config/boolean-config.js]: The description is user supplied so we would need to do some kind of escaping for anything that is not an anchor. I just tried this with a local charm and it executes javascript, renders html etc.

Hopefully there is a lib or something that can help with this.

Comment (hatched) [inline jujugui/static/gui/src/app/components/boolean-config/boolean-config.js]: After putting some thought into the best approach for this here is what I've come up with:

1. Take the description and split on the links ie) <a></a>.
2. Assign each section to an array but instead of adding the strings for the <a>'s, instead, add a jsx <a> using the values provided in the description (see the regex from below).
3. Add this array to the content of the description div.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/4047/
Test PASSed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/4048/
Test PASSed.

File: jujugui/static/gui/src/app/views/utils.js [hunk 4]
```diff
@@ -1812,6 +1812,36 @@ YUI.add('juju-view-utils', function(Y) {
...
+  /**
+    Convert plain text links to anchor tags.
+
+    @method _linkify
+    @param {String} str a string that may have a plain text link or URL in it
+    @returns {String} the string with HTML anchor tags for the link
+  */
+  utils.linkify = function(str) {
+    // Sanitize any malicious HTML or Javascript.
```
Comment (kadams54) [inline jujugui/static/gui/src/app/views/utils.js]: Added this bit here for XSS protection, as well as the bit a few lines down where it encodes any double quotes for the href attribute.

**Comment by kadams54** [comment]
My apologies: I rebased without thinking, so I've noted the places that have changed to address the comments above.

File: jujugui/static/gui/src/test/test_utils.js [hunk 5]
```diff
@@ -1163,4 +1163,69 @@ describe('utilities', function() {
     });
   });
 
+  describe('linkify', function() {
```
Comment (kadams54) [inline jujugui/static/gui/src/test/test_utils.js]: Added a slew of tests to check various URLs, from the good to the bad to the ugly.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/4078/
Test PASSed.

**Comment by hatched** [comment]
:+1: QA :ok: 

After attempting to run common and not so common injection attempts through it didn't allow any of them to be executable.

**Comment by kadams54** [comment]
Thanks all! :shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge
